File: Leetcode/226.cpp
/**
 * 226. 翻转二叉树 BFS
 * link: https://leetcode-cn.com/problems/invert-binary-tree/
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */


class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        queue<TreeNode*> nodelist;
        nodelist.push(root);
        //Pre-order traverse each non-empty node and exchange left and right nodes (subtrees)
        while(!nodelist.empty()){
            TreeNode *temp = nodelist.front();
            nodelist.pop();
            if(temp == NULL)continue;
            TreeNode *temp2 = temp->left;
            temp->left = temp->right;
            temp->right = temp2;
            nodelist.push(temp->left);
            nodelist.push(temp->right);
        }
        return root;
    }
};

File: Leetcode/226.py
# 226. 翻转二叉树 BFS
# link: https://leetcode-cn.com/problems/invert-binary-tree/


class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if not root:
            return None

        nodes = collections.deque([root])
        while nodes:
            node: TreeNode = nodes.popleft()
            node.left, node.right = node.right, node.left
            if node.left:
                nodes.append(node.left)
            if node.right:
                nodes.append(node.right)

        return root
